JAR fails to start with the error "Truffle could not be initialized because Multi-Release classes are not configured correctly"

I am trying to run the GraalPy Micronaut demo (https://github.com/sachin-pikle/graalpy-mn-vader-v2) as a JAR, created using `./mvnw package`, with **GraalVM 25** but it fails with the following error:

`Caused by: java.lang.InternalError: Truffle could not be initialized because Multi-Release classes are not configured correctly. This most likely means Truffle classes have been repackaged incorrectly and the `Multi-Release: true` attribute in META-INF/MANIFEST.MF has been lost. A common cause of this error is invalid Uber JAR configuration. For more information see: https://www.graalvm.org/latest/reference-manual/embed-languages/#uber-jar-file-creation. This check may be disabled with '-Dpolyglotimpl.DisableMultiReleaseCheck=true'.`

Steps to reproduce:
1. Git clone https://github.com/sachin-pikle/graalpy-mn-vader-v2.git
2. `sdk use java 25.0.2-graal`
3. Comment lines 120-153 of the pom.xml (the shade plugin section).
4. `./mvnw package`
5. `java -jar target/graalpy-mn-vader-v2-0.1.jar`
6. You should something similar to the following output with the error.
Server output logs:
```
% java -jar target/graalpy-mn-vader-v2-0.1.jar
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
11:47:17.495 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [io.micronaut.graal.graalpy.GraalPyContext] could not be loaded: Error instantiating bean of type  [io.micronaut.graal.graalpy.GraalPyContext]

Path Taken:
new @i.m.c.a.Context i.m.g.g.GraalPyContext(GraalPyContextBuilderFactory builder)
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [io.micronaut.graal.graalpy.GraalPyContext] could not be loaded: Error instantiating bean of type  [io.micronaut.graal.graalpy.GraalPyContext]

Path Taken:
new @i.m.c.a.Context i.m.g.g.GraalPyContext(GraalPyContextBuilderFactory builder)
        at io.micronaut.context.DefaultBeanContext.initializeEagerBeans(DefaultBeanContext.java:1901)
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1823)
        at io.micronaut.context.DefaultBeanContext.configureAndStartContext(DefaultBeanContext.java:3011)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:330)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:293)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:77)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:354)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:340)
        at graalpy.demo.Application.main(Application.java:8)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [io.micronaut.graal.graalpy.GraalPyContext]

Path Taken:
new @i.m.c.a.Context i.m.g.g.GraalPyContext(GraalPyContextBuilderFactory builder)
        at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2108)
        at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:2803)
        at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:78)
        at io.micronaut.context.DefaultBeanContext.intializeEagerBean(DefaultBeanContext.java:2697)
        at io.micronaut.context.DefaultBeanContext.initializeEagerBean(DefaultBeanContext.java:2345)
        at io.micronaut.context.DefaultBeanContext.initializeEagerBeans(DefaultBeanContext.java:1895)
        ... 8 common frames omitted
Caused by: java.lang.ExceptionInInitializerError
        at org.graalvm.polyglot.io.FileSystem.newDefaultFileSystem(FileSystem.java:591)
        at org.graalvm.python.embedding.VirtualFileSystemImpl.<init>(VirtualFileSystemImpl.java:327)
        at org.graalvm.python.embedding.VirtualFileSystem.<init>(VirtualFileSystem.java:313)
        at org.graalvm.python.embedding.VirtualFileSystem$Builder.build(VirtualFileSystem.java:295)
        at org.graalvm.python.embedding.VirtualFileSystem.create(VirtualFileSystem.java:337)
        at org.graalvm.python.embedding.GraalPyResources.contextBuilder(GraalPyResources.java:265)
        at io.micronaut.graal.graalpy.DefaultGraalPyContextBuilderFactory.createBuilder(DefaultGraalPyContextBuilderFactory.java:28)
        at io.micronaut.graal.graalpy.GraalPyContext.<init>(GraalPyContext.java:38)
        at io.micronaut.graal.graalpy.$GraalPyContext$Definition.instantiate(Unknown Source)
        at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2092)
        ... 13 common frames omitted
Caused by: java.lang.IllegalStateException: Failed to initialize execution listener class.
        at org.graalvm.polyglot.io.IOHelper$ImplHolder.initImpl(IOHelper.java:201)
        at org.graalvm.polyglot.io.IOHelper$ImplHolder.<clinit>(IOHelper.java:192)
        ... 23 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at org.graalvm.polyglot.io.IOHelper$ImplHolder.initImpl(IOHelper.java:198)
        ... 24 common frames omitted
Caused by: java.lang.InternalError: Truffle could not be initialized because Multi-Release classes are not configured correctly. This most likely means Truffle classes have been repackaged incorrectly and the `Multi-Release: true` attribute in META-INF/MANIFEST.MF has been lost. A common cause of this error is invalid Uber JAR configuration. For more information see: https://www.graalvm.org/latest/reference-manual/embed-languages/#uber-jar-file-creation. This check may be disabled with '-Dpolyglotimpl.DisableMultiReleaseCheck=true'.
        at com.oracle.truffle.api.Truffle.createRuntime(Truffle.java:110)
        at com.oracle.truffle.api.Truffle$1.run(Truffle.java:185)
        at com.oracle.truffle.api.Truffle$1.run(Truffle.java:183)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:74)
        at com.oracle.truffle.api.Truffle.initRuntime(Truffle.java:183)
        at com.oracle.truffle.api.Truffle.<clinit>(Truffle.java:64)
        at com.oracle.truffle.api.impl.Accessor.getTVMCI(Accessor.java:1656)
        at com.oracle.truffle.api.impl.Accessor$Constants.<clinit>(Accessor.java:1515)
        at com.oracle.truffle.api.impl.Accessor.engineSupport(Accessor.java:1581)
        at com.oracle.truffle.api.library.LibraryAccessor.engineAccessor(LibraryAccessor.java:57)
        at com.oracle.truffle.api.library.LibraryFactory.loadExternalDefaultProviders(LibraryFactory.java:439)
        at com.oracle.truffle.api.library.LibraryFactory.getExternalDefaultProviders(LibraryFactory.java:430)
        at com.oracle.truffle.api.library.LibraryFactory.initDefaultExports(LibraryFactory.java:213)
        at com.oracle.truffle.api.library.LibraryFactory.<init>(LibraryFactory.java:208)
        at com.oracle.truffle.api.library.DynamicDispatchLibraryGen.<init>(DynamicDispatchLibraryGen.java:33)
        at com.oracle.truffle.api.library.DynamicDispatchLibraryGen.<clinit>(DynamicDispatchLibraryGen.java:25)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:547)
        at com.oracle.truffle.api.library.LibraryFactory.loadGeneratedClass(LibraryFactory.java:775)
        at com.oracle.truffle.api.library.LibraryFactory.resolveImpl(LibraryFactory.java:756)
        at com.oracle.truffle.api.library.LibraryFactory.resolve(LibraryFactory.java:749)
        at com.oracle.truffle.api.library.LibraryFactory.<init>(LibraryFactory.java:202)
        at com.oracle.truffle.api.interop.InteropLibraryGen.<init>(InteropLibraryGen.java:180)
        at com.oracle.truffle.api.interop.InteropLibraryGen.<clinit>(InteropLibraryGen.java:171)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:547)
        at com.oracle.truffle.api.library.LibraryFactory.loadGeneratedClass(LibraryFactory.java:775)
        at com.oracle.truffle.api.library.LibraryFactory.resolveImpl(LibraryFactory.java:756)
        at com.oracle.truffle.api.library.LibraryFactory.resolve(LibraryFactory.java:749)
        at com.oracle.truffle.api.interop.InteropLibrary.<clinit>(InteropLibrary.java:2972)
        at com.oracle.truffle.polyglot.PolyglotValueDispatch.<clinit>(PolyglotValueDispatch.java:168)
        at com.oracle.truffle.polyglot.PolyglotImpl.initialize(PolyglotImpl.java:219)
        at org.graalvm.polyglot.Engine.loadAndValidateProviders(Engine.java:1890)
        at org.graalvm.polyglot.Engine$1.run(Engine.java:1916)
        at org.graalvm.polyglot.Engine$1.run(Engine.java:1911)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:74)
        at org.graalvm.polyglot.Engine.initEngineImpl(Engine.java:1911)
        at org.graalvm.polyglot.Engine$ImplHolder.<clinit>(Engine.java:179)
        at org.graalvm.polyglot.Engine.getImpl(Engine.java:518)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        ... 26 common frames omitted
```

To resolve the issue, I had to add the following snippet to my pom.xml based on the GraalVM docs (https://www.graalvm.org/latest/reference-manual/embed-languages/#uber-jar-file-creation) and AI assistance:

```
      <!-- end::graalpy-maven-plugin[] -->
      <!-- GraalVM Embedding Uber-JAR Shade Plugin -->
      <!-- Added to fix java.lang.InternalError: Truffle could not be initialized because Multi-Release classes are not configured correctly.-->
      <!-- This most likely means Truffle classes have been repackaged incorrectly and the `Multi-Release: true` attribute in META-INF/MANIFEST.MF has been lost. -->
      <!-- A common cause of this error is invalid Uber JAR configuration. -->
      <!-- For more information see: https://www.graalvm.org/latest/reference-manual/embed-languages/#uber-jar-file-creation -->
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-shade-plugin</artifactId>
        <version>3.6.0</version>
        <executions>
          <execution>
            <phase>package</phase>
            <goals>
              <goal>shade</goal>
            </goals>
            <configuration>
              <transformers>
                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
                <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
                  <mainClass>${exec.mainClass}</mainClass>
                  <manifestEntries>
                    <Multi-Release>true</Multi-Release>
                  </manifestEntries>
                </transformer>
              </transformers>
              <filters>
                <filter>
                  <artifact>*:*:*:*</artifact>
                  <excludes>
                    <exclude>META-INF/*.SF</exclude>
                    <exclude>META-INF/*.DSA</exclude>
                    <exclude>META-INF/*.RSA</exclude>
                  </excludes>
                </filter>
              </filters>
            </configuration>
          </execution>
        </executions>
      </plugin>
      <!-- end: GraalVM Embedding Uber-JAR Shade Plugin -->
```

--- a/micronaut-maven-plugin/src/site/asciidoc/examples/package.adoc
+++ b/micronaut-maven-plugin/src/site/asciidoc/examples/package.adoc
@@ -77,6 +77,54 @@ The plugin does not currently provide a dedicated non-shaded launcher for plain 
 depends on separate classpath entries, prefer `docker` packaging or your own classpath-based launch script over the
 default shaded JAR.
 
+If you intentionally keep a shaded fat JAR for a GraalVM Truffle or GraalPy application, follow the
+https://www.graalvm.org/latest/reference-manual/embed-languages/#uber-jar-file-creation[GraalVM uber-JAR guidance] and
+make the application POM preserve the multi-release manifest metadata required by Truffle. Configure the Shade plugin
+with both the services transformer and a manifest transformer that keeps `${exec.mainClass}` as the main class and adds
+`Multi-Release: true`:
+
+[source,xml]
+----
+<plugin>
+  <groupId>org.apache.maven.plugins</groupId>
+  <artifactId>maven-shade-plugin</artifactId>
+  <executions>
+    <execution>
+      <phase>package</phase>
+      <goals>
+        <goal>shade</goal>
+      </goals>
+      <configuration>
+        <transformers>
+          <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+          <transformer implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+            <mainClass>${exec.mainClass}</mainClass>
+            <manifestEntries>
+              <Multi-Release>true</Multi-Release>
+            </manifestEntries>
+          </transformer>
+        </transformers>
+        <filters>
+          <filter>
+            <artifact>*:*</artifact>
+            <excludes>
+              <exclude>META-INF/*.SF</exclude>
+              <exclude>META-INF/*.DSA</exclude>
+              <exclude>META-INF/*.RSA</exclude>
+            </excludes>
+          </filter>
+        </filters>
+      </configuration>
+    </execution>
+  </executions>
+</plugin>
+----
+
+Micronaut Maven Plugin does not add this manifest entry for every application because most applications do not need
+Truffle's multi-release metadata. To verify the packaging, inspect `META-INF/MANIFEST.MF` in the generated JAR and
+confirm it contains `Multi-Release: true`, then start the application with the same JDK used in production. Do not
+disable Truffle's multi-release validation check as a packaging fix.
+
 === Generating GraalVM native images
 
 ----

